### PR TITLE
Write the stress test for gRPC, http, and tcp probes

### LINKED ISSUE #115782: Write the stress test for gRPC, http, and tcp probes
Write a new e2e test that will be marked as `[Serial]` and will be covering this bug fix: https://github.com/kubernetes/kubernetes/issues/89898

The test will serve as a replacement for the skipped unit test: https://github.com/kubernetes/kubernetes/pull/115329

The logic of the test will be as following:

- Create MANY containers (see the unit test logic)
- Each container has a liveness probe with `1s` interval
- After some time validate that neither of containers has `restartCount > 0`

/cc @aojea 
/sig node
/kind cleanup
/priority important-soon
/area tests
/help-wanted
/good-first-issue

This is part of KEP: https://github.com/kubernetes/enhancements/issues/2727 GA requirements, but also need the same tests for http and tcp.

### PR DESCRIPTION
**Pull request by lanycrost**
#### What type of PR is this?
/kind cleanup
/area tests

@kubernetes/sig-architecture-pr-reviews @kubernetes/sig-node-pr-reviews

#### What this PR does / why we need it:

This PR adds stress tests for GRPC, HTTP and TCP probes with this logic.

- Create MANY containers.
- Each container has a liveness probe with 1s interval.
- After some time validate that neither of containers has restartCount > 0.

#### Which issue(s) this PR fixes:

Fixes #115782

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/node/stress_probe.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	numContainers                = 100
+	defaultObservationTimeout    = time.Minute * 4
+)
+
+var _ = SIGDescribe("Stress test probes", func() {
+	f := framework.NewDefaultFramework("stress-test-probe")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelBaseline
+
+	/*
+		Release: v1.23
```
Comment (bart0sh) [inline test/e2e/node/stress_probe.go:47]: Why Release 1.23?

Comment (lanycrost) [reply test/e2e/node/stress_probe.go:47]: I just didn't know what to add there, I can leave it blank, or change it.

File: test/e2e/node/stress_probe.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+)
+
+const (
+	probeTestInitialDelaySeconds = 15
+	numTestPods                  = 1
+	numContainers                = 100
+	defaultObservationTimeout    = time.Minute * 4
+)
```
Comment (yuzhiquan) [inline test/e2e/node/stress_probe.go:40]: How about
```suggestion
containerPort = 2000
```

File: test/e2e/node/stress_probe.go (hunk 1 above)
Comment (bart0sh) [reply test/e2e/node/stress_probe.go:47]: Set it to 1.28 unless you want to backport this to previous releases.

File: test/e2e/node/stress_probe.go (hunk 2 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go:40]: You mean use this constant instead declarations across whole code?

Comment (bart0sh) [reply test/e2e/node/stress_probe.go:40]: yes. At least I understood it this way.

File: test/e2e/node/stress_probe.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+)
+
+const (
+	probeTestInitialDelaySeconds = 15
+	numTestPods                  = 1
+	numContainers                = 100
```
Comment (liggitt) [inline test/e2e/node/stress_probe.go:38]: do we normally put stress tests in e2e tests or in scale tests? I'm a little worried that not all clusters running e2e tests run on machines that expect/handle high load

File: test/e2e/node/stress_probe.go [hunk 4]
```diff
@@ -1,9 +1,12 @@
 /*
-Copyright 2023 The Kubernetes Authors.
+Copyright 2015 The Kubernetes Authors.
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:2]: why this change?

File: test/e2e/node/stress_probe.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+const (
+	probeTestInitialDelaySeconds = 15
+	numTestPods                  = 1
+	numContainers                = 100
+	defaultObservationTimeout    = time.Minute * 4
+	containerPort                = 2000
+)
+
+var _ = SIGDescribe("Stress test probes", func() {
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:46]:
```suggestion
var _ = SIGDescribe("[Serial] [Slow] [Disruptive] Stress test probes", func() {
```

File: test/e2e/node/stress_probe.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	})
+})
+
+func livenessPodSpec(probeType, namespace string, livenessProbe v1.Probe, ports []v1.ContainerPort, args ...string) *v1.Pod {
+	pod := e2epod.NewAgnhostPod(namespace, "liveness-stress-"+string(uuid.NewUUID()), nil, nil, ports, args...)
+	pod.ObjectMeta.Labels = map[string]string{"test": "liveness-stress"}
+	pod.Spec.Containers[0].LivenessProbe = &livenessProbe
+	pod.Spec.Containers[0].Name = "agnhost-container-0"
+	pod.Spec.Containers = append(pod.Spec.Containers, make([]v1.Container, numContainers-1)...)
+	for idx := range make([]int, numContainers) {
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:114]: why do you need this make here, is not just a `for idx:=0; idx < numContainers ; idx++`

File: test/e2e/node/stress_probe.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+			pod.Spec.Containers[idx].Args[3] = strconv.Itoa(newContainerPort)
+		case "http":
+			pod.Spec.Containers[idx].LivenessProbe.HTTPGet.Port = intstr.FromInt(newContainerPort)
+			pod.Spec.Containers[idx].Args[3] = strconv.Itoa(newContainerPort)
+		case "grpc":
+			pod.Spec.Containers[idx].LivenessProbe.GRPC.Port = int32(newContainerPort)
+			pod.Spec.Containers[idx].Args[2] = strconv.Itoa(newContainerPort)
+			pod.Spec.Containers[idx].Args[4] = strconv.Itoa(newContainerPort + containerPort)
+		}
+	}
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:131]: you better create a generator and call it in each iteration, is easier to read
 `newContainer(name string, probeType string) *v1.Container`

File: test/e2e/node/stress_probe.go [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+func tcpSocketHandler(port int) v1.ProbeHandler {
+	return v1.ProbeHandler{
+		TCPSocket: &v1.TCPSocketAction{
+			Port: intstr.FromInt(port),
+		},
+	}
+}
+
+// RunLivenessTest verifies the number of restarts.
+func RunLivenessTest(ctx context.Context, f *framework.Framework, pod *v1.Pod, timeout time.Duration) {
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:153]: why export it? should it be private `runLivenessTest`

File: test/e2e/node/stress_probe.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+		framework.ExpectNoError(err, fmt.Sprintf("getting pod %s", pod.Name))
+		for idx := range [numContainers]int{} {
+			containerName := "agnhost-container-" + strconv.Itoa(idx)
+
+			restartCount := podutil.GetExistingContainerStatus(pod.Status.ContainerStatuses, containerName).RestartCount
+
+			if restartCount != int32(0) {
+				framework.Failf("Restart count of pod's container %s/%s/%s is now %d (%v elapsed)",
+					ns, pod.Name, containerName, restartCount, time.Since(start))
+				break cron
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:190]: avoid loops if possible, you can achieve the same with different logic

File: test/e2e/node/stress_probe.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+		fmt.Sprintf("starting pod %s in namespace %s", pod.Name, ns))
+	framework.Logf("Started pod %s in namespace %s", pod.Name, ns)
+
+	// Check the pod's current state and verify that restartCount is present.
+	ginkgo.By("checking the pod's current state and verifying that restartCount is present")
+	pod, err := podClient.Get(ctx, pod.Name, metav1.GetOptions{})
+	framework.ExpectNoError(err, fmt.Sprintf("getting pod %s in namespace %s", pod.Name, ns))
+
+cron:
+	for start := time.Now(); time.Now().Before(deadline); time.Sleep(60 * time.Second) {
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:179]: why this sleep?

File: test/e2e/node/stress_probe.go [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+	return v1.ProbeHandler{
+		TCPSocket: &v1.TCPSocketAction{
+			Port: intstr.FromInt(port),
+		},
+	}
+}
+
+// RunLivenessTest verifies the number of restarts.
+func RunLivenessTest(ctx context.Context, f *framework.Framework, pod *v1.Pod, timeout time.Duration) {
+	deadline := time.Now().Add(timeout)
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:154]: you should not take into account the time to create pod since is not deterministic

**Comment by bart0sh** [comment]
@lanycrost please address review comments, thanks

**Comment by lanycrost** [comment]
@aojea please review code again, I'm thinking about adding some kind of idle interval to wait and understand may be pod probes will start failing after some time, that's why I've added the loop with time.

**Comment by lanycrost** [comment]
@aojea  as well may be will be better to increase number of containers?

File: test/e2e/node/stress_probe.go [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+	pod.Spec.Containers = append(pod.Spec.Containers, make([]v1.Container, numContainers-1)...)
+	for idx := 0; idx < numContainers; idx++ {
+		name := "agnhost-container-" + strconv.Itoa(idx+1)
+		instance := *pod.Spec.Containers[0].DeepCopy()
+		pod.Spec.Containers[idx] = newContainer(name, probeType, idx, instance)
+	}
+	return pod
+}
+
+func newContainer(name, probeType string, containerIdx int, instance v1.Container) v1.Container {
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:119]: this function is super confusing because it works based on the knowledge of the `probeArgs` and where exactly has to fill each parameter

File: test/e2e/node/stress_probe.go [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+					Port:    containerPort,
+					Service: nil,
+				},
+			},
+			InitialDelaySeconds: probeTestInitialDelaySeconds,
+			TimeoutSeconds:      5, // default 1s can be pretty aggressive in CI environments with low resources
+			FailureThreshold:    1,
+		}
+		ports := []v1.ContainerPort{{ContainerPort: int32(containerPort)}}
+		probeArgs := []string{"grpc-health-checking", "--port", "", "--http-port", ""}
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:99]: I think that you can just add it entirely in `newContainer` based on the type, can't you?

This logic is very brittle since it depends on the positional arguments and it goes through two other functions later that fill in the gaps, there has to be a more readable way ,  can't we just add the entire command/args inside `newContainer`? is there any reason why we have to declare them here?

File: test/e2e/node/stress_probe.go [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+func livenessPodSpec(probeType, namespace string, livenessProbe v1.Probe, ports []v1.ContainerPort, args ...string) *v1.Pod {
+	pod := e2epod.NewAgnhostPod(namespace, "liveness-stress-"+string(uuid.NewUUID()), nil, nil, ports, args...)
+	pod.ObjectMeta.Labels = map[string]string{"test": "liveness-stress"}
+	pod.Spec.Containers[0].LivenessProbe = &livenessProbe
+	pod.Spec.Containers[0].Name = "agnhost-container-0"
+	pod.Spec.Containers = append(pod.Spec.Containers, make([]v1.Container, numContainers-1)...)
+	for idx := 0; idx < numContainers; idx++ {
+		name := "agnhost-container-" + strconv.Itoa(idx+1)
+		instance := *pod.Spec.Containers[0].DeepCopy()
+		pod.Spec.Containers[idx] = newContainer(name, probeType, idx, instance)
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:114]: you override container[0] here, is that on purpose?

File: test/e2e/node/stress_probe.go [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+		ginkgo.By("deleting the pod")
+		return podClient.Delete(ctx, pod.Name, *metav1.NewDeleteOptions(0))
+	})
+	ginkgo.By(fmt.Sprintf("Creating pod %s in namespace %s", pod.Name, ns))
+	podClient.Create(ctx, pod)
+
+	// Wait until the pod is not pending. (Here we need to check for something other than
+	// 'Pending' other than checking for 'Running', since when failures occur, we go to
+	// 'Terminated' which can cause indefinite blocking.)
+	framework.ExpectNoError(e2epod.WaitForPodNotPending(ctx, f.ClientSet, ns, pod.Name),
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:181]: should not we wait for pod running and read?

File: test/e2e/node/stress_probe.go [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+		fmt.Sprintf("starting pod %s in namespace %s", pod.Name, ns))
+	framework.Logf("Started pod %s in namespace %s", pod.Name, ns)
+
+	// Check the pod's current state and verify that restartCount is present.
+	ginkgo.By("checking the pod's current state and verifying that restartCount is present")
+	pod, err := podClient.Get(ctx, pod.Name, metav1.GetOptions{})
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:187]: the problem with the probes is that they accumulate connections, we should check X times during Y minutes/seconds, at least 2 mins

File: test/e2e/node/stress_probe.go (hunk 12 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go:119]: what you may suggest here? What parameters will be better to pass here. I'm passing probeType because it's hard to understand what type of probes is enabled, will apriciate if you have suggestions for this part as well.

File: test/e2e/node/stress_probe.go (hunk 13 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go:99]: yeah, I will try to change logic here.

File: test/e2e/node/stress_probe.go (hunk 14 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go:114]: I'm override here to have similar names for all containers inside the pod.

File: test/e2e/node/stress_probe.go (hunk 16 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go:187]: yeah, I just removed that part, I will add it back

**Comment by lanycrost** [comment]
@aojea I've made few changes, waiting for your feedback

**Comment by aojea** [comment]
/test

can we run it on presubmits?

**Comment by aojea** [comment]
overall lgtm, maybe specify the timeouts to make it visible on the tests, but I defer to node people and to see if we can have some signal in presubmit

/assign @bobbypage @SergeyKanzhelev

File: test/e2e/node/stress_probe.go [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+	defaultObservationTimeout    = time.Minute * 4
+)
+
+type containerConfig struct {
+	Name         string
+	ProbeType    string
+	ContainerIdx int
+}
+
+var _ = SIGDescribe("[Serial] [Slow] [Disruptive] Stress test probes", func() {
```
Comment (aojea) [inline test/e2e/node/stress_probe.go:51]: @rphillips we asked to add all these tags to indicate these tests are not meant to run by default, you can see that it has `Disruptive` already

**Comment by lanycrost** [comment]
@bobbypage @SergeyKanzhelev What you think, do we need anything to add here?

**Comment by bart0sh** [comment]
@lanycrost I didn't review this PR, but I'd suggest to fix failing test if it's caused by your changes.

File: test/e2e/node/stress_probe.go [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+	f := framework.NewDefaultFramework("stress-test-probe")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelBaseline
+
+	/*
+		Release: v1.28
```
Comment (bart0sh) [inline test/e2e/node/stress_probe.go:56]: This should be updated to 1.30 I guess.

File: test/e2e/node/stress_probe.go [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+		count += int(containerStatus.RestartCount)
+	}
+	return count
+}
+
+// runLivenessTest verifies the number of restarts.
+func runLivenessTest(ctx context.Context, f *framework.Framework, pod *v1.Pod) {
+	podClient := e2epod.NewPodClient(f)
+	ns := f.Namespace.Name
+	gomega.Expect(pod.Spec.Containers).NotTo(gomega.BeEmpty())
```
Comment (bart0sh) [inline test/e2e/node/stress_probe.go:173]: I'm not sure we need to check anything here, but if we do it probably makes sense to expect pod to have `NumContainers` containers.

File: test/e2e/node/stress_probe.go [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+	ginkgo.By(fmt.Sprintf("Creating pod %s in namespace %s", pod.Name, ns))
+	podClient.Create(ctx, pod)
+
+	// Wait until the pod is running.
+	framework.ExpectNoError(e2epod.WaitForPodRunningInNamespace(ctx, f.ClientSet, pod),
+		fmt.Sprintf("starting pod %s in namespace %s", pod.Name, ns))
+	framework.Logf("Started pod %s in namespace %s", pod.Name, ns)
+
+	deadline := time.Now().Add(defaultObservationTimeout)
+	for start := time.Now(); time.Now().Before(deadline); time.Sleep(10 * time.Second) {
```
Comment (bart0sh) [inline test/e2e/node/stress_probe.go:188]: Would it be more clear if we simply run this loop `defaultObservationTimeout  / 10` times?

File: test/e2e/node/stress_probe.go [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+)
+
+const (
+	probeTestInitialDelaySeconds = 15
```
Comment (bart0sh) [inline test/e2e/node/stress_probe.go:38]: This constant, `defaultObservationTimeout` and some functions(`getRestartCount` at least) are the same as in the test/e2e/common/node/container_probe.go, should it be shared somehow?

File: test/e2e/node/stress_probe.go (hunk 18 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go:56]: OK will update it

File: test/e2e/node/stress_probe.go (hunk 21 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go:38]: I moved them to framework

File: test/e2e/node/stress_probe.go (hunk 20 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go]: changed, please check

File: test/e2e/node/stress_probe.go (hunk 19 above)
Comment (lanycrost) [reply test/e2e/node/stress_probe.go]: changed, please check this one too.

File: test/e2e/framework/framework.go [hunk 22]
```diff
@@ -56,7 +56,9 @@ import (
 const (
 	// DefaultNamespaceDeletionTimeout is timeout duration for waiting for a namespace deletion.
 	DefaultNamespaceDeletionTimeout = 5 * time.Minute
-	defaultServiceAccountName       = "default"
+	// DefaultObservationTimeout is duration where checking kubernetes object status.
```
Comment (reinka) [inline test/e2e/framework/framework.go:59]: There might be a little typo in `is duration where checking`, maybe this should mean something along the lines of
`// DefaultObservationTimeout specifies the maximum duration to wait when checking the status of kubernetes objects`?

**Comment by lanycrost** [comment]
Are we going to merge this pull request?

@bart0sh @SergeyKanzhelev @pacoxu @bart0sh

File: test/e2e/node/stress_probe.go (hunk 3 above)
Comment (pacoxu) [reply test/e2e/node/stress_probe.go]: https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/test/e2e/windows/density.go 

It is similar to the density test or kubelet perf.

https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/test/e2e/node/kubelet_perf.go#L199-L200
